ui/src/components/VirtualKeyboard.tsx

```diff
@@ -112,78 +129,123 @@ function KeyboardWrapper() {
+    if (key === "CtrlAltBackspace") {
+      sendKeyboardEvent(
+        [keys["Backspace"]],
+        [modifiers["ControlLeft"], modifiers["AltLeft"]],
+      );
 
-        if (isCapsLockActive) {
-          setIsCapsLockActive(false);
-          sendKeyboardEvent([keys["CapsLock"]], []);
-          return;
-        }
-      }
+      setTimeout(resetKeyboardState, 100);
+      return;
+    }
 
-      // Handle caps lock state change
-      if (isKeyCaps) {
-        setIsCapsLockActive(!isCapsLockActive);
-      }
+    // strip away the parens for shifted characters
+    const cleanKey = key.replace(/[()]/g, "");
+
+    // if it's one of the special keys, just pass it through immediately and return
+    const passthrough = ["PrintScreen", "SystemRequest", "Pause", "Break", "ScrollLock", "Enter", "Space"].find((value) => value === cleanKey);
+
+    if (passthrough) {
+      emitkeycode(cleanKey);
+      return;
+    }
+
+    // adjust the sticky state of the Shift/Ctrl/Alt/Meta/AltGr
+    if (key === "CapsLock")
+      setIsCapsLockActive(!isCapsLockActive);
+    else if (key === "ShiftLeft" || key === "ShiftRight")
+      setIsShiftActive(!isShiftActive);
+    else if (key === "ControlLeft" || key === "ControlRight")
+      setIsCtrlActive(!isCtrlActive);
+    else if (key === "AltLeft" || key === "AltRight")
+      setIsAltActive(!isAltActive);
+    else if (key === "MetaLeft" || key === "MetaRight")
+      setIsMetaActive(!isMetaActive);
+    else if (key === "AltGr")
+      setIsAltGrActive(!isAltGrActive);
+
+    emitkeycode(cleanKey);
+
+    function emitkeycode(key: string) {
+      const effectiveMods: number[] = [];
 
-      // Collect new active keys and modifiers
-      const newKeys = keys[cleanKey] ? [keys[cleanKey]] : [];
-      const newModifiers =
-        keyHasShiftModifier && !isCapsLockActive ? [modifiers["ShiftLeft"]] : [];
+      if (isShiftActive)
+        effectiveMods.push(modifiers["ShiftLeft"]);
 
-      // Update current keys and modifiers
-      sendKeyboardEvent(newKeys, newModifiers);
+      if (isCtrlActive)
+        effectiveMods.push(modifiers["ControlLeft"]);
 
-      // If shift was used as a modifier and caps lock is not active, revert to default layout
-      if (keyHasShiftModifier && !isCapsLockActive) {
-        setLayoutName("default");
+      if (isAltActive)
+        effectiveMods.push(modifiers["AltLeft"]);
+
+      if (isMetaActive)
+        effectiveMods.push(modifiers["MetaLeft"]);
+
+      if (isAltGrActive) {
+        effectiveMods.push(modifiers["MetaRight"]);
+        effectiveMods.push(modifiers["CtrlLeft"]);
       }
 
-      setTimeout(resetKeyboardState, 100);
-    },
-    [isCapsLockActive, sendKeyboardEvent, resetKeyboardState, setIsCapsLockActive],
+      const keycode = keys[key];
+      if (keycode) {
+        // send the keycode with modifiers
+        sendKeyboardEvent([keycode], effectiveMods);
+      }
+
+      // release the key (if one pressed), but retain the modifiers
```

Comment:
Here's the biggest gain... we can leave the modifier state still reflecting what we're tracking it to be so that the remote knows we still have the corresponding keys down. This is especially important for `CapsLock` because we want the remote to know we've got that on so it can do the warnings and such.